ddtrace/internal/ci_visibility/recorder.py

```diff
@@ -0,0 +1,124 @@
+import json
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+import ddtrace
+from ddtrace import Tracer
+from ddtrace.contrib import trace_utils
+from ddtrace.ext import ci
+from ddtrace.ext import test
+from ddtrace.internal import atexit
+from ddtrace.internal.ci_visibility.filters import TraceCiVisibilityFilter
+from ddtrace.internal.compat import parse
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.service import Service
+from ddtrace.settings import IntegrationConfig
+
+
+log = get_logger(__name__)
+
+
+def _extract_repository_name_from_url(repository_url):
+    # type: (str) -> str
+    try:
+        return parse.urlparse(repository_url).path.rstrip(".git").rpartition("/")[-1]
+    except ValueError:
+        # In case of parsing error, default to repository url
+        log.warning("Repository name cannot be parsed from repository_url: %s", repository_url)
+        return repository_url
+
+
+class CIVisibility(Service):
+    _instance = None  # type: Optional[CIVisibility]
+    enabled = False
+
+    def __init__(self, tracer=None, config=None, service=None):
+        # type: (Optional[Tracer], Optional[IntegrationConfig], Optional[str]) -> None
+        super(CIVisibility, self).__init__()
+
+        self.tracer = tracer or ddtrace.tracer
+        self.config = config  # type: Optional[IntegrationConfig]
+        self._tags = ci.tags()  # type: Dict[str, str]
+        self._service = service
+        self._codeowners = None
+
+        int_service = None
+        if self.config is not None:
+            int_service = trace_utils.int_service(None, self.config)
+        # check if repository URL detected from environment or .git, and service name unchanged
+        if self._tags.get(ci.git.REPOSITORY_URL, None) and self.config and int_service == self.config._default_service:
+            self._service = _extract_repository_name_from_url(self._tags[ci.git.REPOSITORY_URL])
+        elif self._service is None and int_service is not None:
+            self._service = int_service
+
+        try:
+            from ddtrace.internal.codeowners import Codeowners
+
+            self._codeowners = Codeowners()
+        except ValueError:
+            log.warning("CODEOWNERS file is not available")
+        except Exception:
+            log.warning("Failed to load CODEOWNERS", exc_info=True)
+
+    @classmethod
+    def enable(cls, tracer=None, config=None, service=None):
+        # type: (Optional[Tracer], Optional[Any], Optional[str]) -> None
+
+        if cls._instance is not None:
+            log.debug("%s already enabled", cls.__name__)
+            return
+        log.debug("Enabling %s", cls.__name__)
+
+        cls._instance = cls(tracer=tracer, config=config, service=service)
+        cls.enabled = True
+
+        cls._instance.start()
+        atexit.register(cls.disable)
+
+        log.debug("%s enabled", cls.__name__)
+
+    @classmethod
+    def disable(cls):
```

Comment:
Do we actually ever need to disable CI Visibility mode once it's enabled?

Reply:
Assuming that CI Visibility will only enabled and run for each CI job, I don't believe we need to disable it. However "disabling" it in this context is just explicitly shutting down the tracer, so this shouldn't matter in the long run.